Add transition path

s8au opened the pull request:
Add CO2 constraints and transition path to endogenous transport. Must be run with snakemake

s8au committed:
Add transition path

Add CO2 constraint and transition path
Save result as png for each year

s8au committed:
Merge remote-tracking branch 'upstream/master'

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -75,26 +76,27 @@ def transport(string):
     oil = df_oil.set_index(['parameter'])
     #discount rate 0.07
-    oil_capital_cost = annuity(oil.at['lifetime','value'], 0.07)*oil.at['investment','value']*1000*+oil.at['FOM','value']/100 # in €/MW
+    oil_capital_cost = oil.at['investment','value']*1000*(annuity(oil.at['lifetime','value'], 0.07)+oil.at['FOM','value']/100) # in €/MW
```
this line can be removed
oil_capital_cost are not used because the oil generator is just an abstraction to represent the supply of oil in the model (I think you understood that and this is just a leftover)

s8au replied:
Should be solved with commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -75,26 +76,27 @@ def transport(string):
     #discount rate 0.07
-    oil_capital_cost = annuity(oil.at['lifetime','value'], 0.07)*oil.at['investment','value']*1000*+oil.at['FOM','value']/100 # in €/MW
+    oil_capital_cost = oil.at['investment','value']*1000*(annuity(oil.at['lifetime','value'], 0.07)+oil.at['FOM','value']/100) # in €/MW
     oil_marginal_cost = oil.at['VOM','value']
```
oil_marginal_cost should be the cost of the fuel, see e.g

https://github.com/PyPSA/pypsa-eur/blob/4c41411f7b01f1a4fc936b652d45d7636cf35b46/scripts/prepare_sector_network.py#L2659

s8au replied:
Should be solved with commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -75,26 +76,27 @@ def transport(string):
     network.add("Bus", 
                 "oil bus", 
-                carrier="oil")
+                carrier="co2_oil")
```
the energy carrier should be oil instead of co2_oil

s8au replied:
Should be solved with commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -75,26 +76,27 @@ def transport(string):
     network.add("Generator",
                 'oil',
                 bus="oil bus",
                 p_nom=100,
```
if you set up p_nom_extendable (capacity can be extended and optimize) do not fix a p_nom (exogenously fixed capacity)

s8au replied:
Should be solved with commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -111,36 +113,92 @@ def transport(string):
     solar = df_tech.loc['solar']
     solar = df_tech.loc['solar'].set_index(['parameter'])
-    capital_cost_solar = annuity(solar.at['lifetime','value'], 0.07)*solar.at['investment','value']*1000*+solar.at['FOM','value']/100 # in €/MW
+    capital_cost_solar = solar.at['investment','value']*1000*(annuity(solar.at['lifetime','value'], 0.07)+solar.at['FOM','value']/100) # in €/MW
```
For the next iteration, could you implement the use of the function
prepare_costs()
https://github.com/PyPSA/pypsa-eur/blob/4c41411f7b01f1a4fc936b652d45d7636cf35b46/scripts/prepare_sector_network.py#L730
This is in preparation for a later implementation into the main pypsa-eur-sec code

s8au replied:
Added a prepare_costs function in commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp replied:
Great, now what is missing is that capital_cost attributed, efficiencies, etc. for every generator is defined via the output of the prepare_costs()
such as in https://github.com/PyPSA/pypsa-eur/blob/80eadcf24aea9b1805d8ebfa66de0422e10f0c04/scripts/prepare_sector_network.py#L776

s8au replied:
Working on it. It is partially implemented in PR #30

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -111,36 +113,92 @@ def transport(string):
+    #set first column as second index and keep original index
+    df_costs.set_index(df_costs.columns[0],append=True,inplace=True)
+    #set parameter/value names as column name, min_count=1 fills not existing values with NaN
+    costs = df_costs.loc[:,"value"].unstack(level=1).groupby("technology").sum(min_count=1)
```
this will be contained in prepare_costs() so that you only need to deal with all costs transformation once

s8au replied:
Moved to function in commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -111,36 +113,92 @@ def transport(string):
                 p_max_pu = CF_solar)
     
-    
+    network.add("Bus", "battery bus")
```
this means to represent the EV battery, right?

Then, name this "EV battery" for compatibility with the main code
https://github.com/PyPSA/pypsa-eur-sec/blob/e3ffb66267c324b259b4a3db2cf03a3a642044c6/scripts/prepare_sector_network.py#L1453

(because there is another battery bus corresponding to static batteries)

s8au replied:
Changed that with commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -111,36 +113,92 @@ def transport(string):
     network.add("Carrier", "vehicle")
     
     network.add("Bus", 
                 "vehicle bus",
```
I suggest we name this "land transport" bus

s8au replied:
Did that with commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -111,36 +113,92 @@ def transport(string):
     
     network.add("Bus", 
                 "vehicle bus",
                 carrier="vehicle")
```
the carriers are energy carriers, they should describe what form of energy is being used in that bus, I suggest naming this "land transport demand"

s8au replied:
Done with commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -111,36 +113,92 @@ def transport(string):
+    #co2_limit = co2_limit * Nyear
+    target = 0#3.392052e+03/8760
     
+    network.add("Store",
+                "co2 atmosphere",
+                e_initial= 0, 
+                e_nom = co2_limit,
+                e_min_pu = -1,
+                bus="co2 atmosphere", 
+                carrier="co2")
```
@ricnogfer check this simple way of setting a CO2 constraint using a store with e_nom=co2_limit

ricnogfer replied:
Ok, I believe you mentioned this to simplify the code implemented some weeks ago for the local/nodal co2 atmosphere found here: https://github.com/ricnogfer/pypsa-eur/blob/local_co2_atmosphere/scripts/prepare_sector_network.py#L598-L606. Will refactor the code accordingly then.

martavp replied:
Yes, I do. 
'e_max_pu' is the maximum filling level per unit (that is, the maximum value is 1)
'e_nom' is the energy capacity of the store, in your case the maximum CO2 that can be stored
I think your former implementation does not work because it implies maximum Co2 stored is 1.

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -75,26 +76,27 @@ def transport(string):
                 'oil',
                 bus="oil bus",
                 p_nom=100,
                 marginal_cost = oil_marginal_cost, #million EUR/MWh
```
the units by default used in PyPSA-Eur-Sec are EUR/MWh (without the million)

s8au replied:
Removed in commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -111,36 +113,92 @@ def transport(string):
     # capital_cost_solar = annuity(25, 0.07)*425000*(1+0.03) # in €/MW
     network.add("Generator",
                 "solar",
                 bus="electricity bus",
```
where is the "electricity bus" added?

s8au replied:
The bus is added in line 107/109

martavp replied:
perfect, I missed it

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -111,36 +113,92 @@ def transport(string):
+    costs = df_costs.loc[:,"value"].unstack(level=1).groupby("technology").sum(min_count=1)
+    network.add("Store", "battery storage",
+                bus="battery bus",
+                e_nom_extendable=True,
```
This should be e_nom_extendable=True (as it is now)
but an additional constraint should be added to relate the
p_nom_opt of the "battery charger" link (i.e. the number of electric cars times the charging capacity of each one) and e_nom_opt (i.e. the number of electric cars times the energy capacity of the battery of each one)

s8au replied:
Added a constraint in commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp replied:
great, the constraints looks correct 
(I'm only looking at the code now, and will run it after reviewing this commit)

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -111,36 +113,92 @@ def transport(string):
+                efficiency = costs.at['battery inverter', 'efficiency']**0.5,
+                capital_cost = costs.at['battery inverter', 'investment']*(annuity(costs.at['battery inverter', 'lifetime'], 0.07)+costs.at['battery inverter', 'FOM']/100),
+                lifetime = costs.at['battery inverter', 'lifetime'])
+    network.add("Link",
```
the "battery discharger" should only be added if the "v2g" (vehicle to grid) option is activated, like in
https://github.com/PyPSA/pypsa-eur-sec/blob/e3ffb66267c324b259b4a3db2cf03a3a642044c6/scripts/prepare_sector_network.py#L1495

s8au replied:
Should be solved with commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -153,7 +211,7 @@ def transport(string):
         bus0="electricity bus",                               
```
there should be two different links "ICE vehicle " and "EV"

s8au replied:
Link with ICE vehicle is from line 137/193 to 148/206
Link with EV starts from line 150/208
you can see it, if you expand the code in between.
Or do you mean something different?

martavp replied:
You are right, I read it to fast, forget about it. I mark this point as resolved.

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -153,7 +211,7 @@ def transport(string):
         bus1 = "vehicle bus",
         carrier = "vehicle",
-        efficiency = 0.85,        
+        efficiency = bev_charge_efficiency,        
```
The "EV" link should connect the "EV battery" with the "land transport demand" bus and its efficiency should be 1
see the reasoning in 
https://github.com/ricnogfer/pypsa-eur-sec_land_transport/issues/9

@BertoGBG I mentioned you here so that you realize this

s8au replied:
Okay, I would then use bev_charge_efficiency for the "battery charger" Link. Or would you keep the general efficiency for the battery inverter from the costs file?

martavp replied:
Yes, let's use the bev_charger_efficiency from the "battery charger" link and read it from the config file.

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -164,36 +222,79 @@ def transport(string):
+        # print solar generator value
+        print("Solar generator value: %f" % network.generators.p_nom_opt["solar"])
+    
+        network.generators_t.p.plot()
```
This output figure is great but try to generate additional figures that help us understand if the endogenous transport is working properly, e.g
[ ] Evolution of the land transport demand supplied with electricity and oil throughout the different planning horizons
[ ] Evolution of the number of EVs and ICE cars throughout the different planning horizons
[ ] Evolution of the filling levels of the EV battery for different planning horizons

s8au replied:
Added plot for generators and ratio of EV and ICEs in commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp replied:
In the output figure 'generators.png'. what are the units on the y axis?

In the output figure 'vehiclenb.png'. what are the units on the y axis? why are the EVs decreasing? shouldn't it be the opposite?

Can you add some plots to verify that the results make sense?

Also, can you add some plots to verify the charging pattern/filling levels of the EVs? E.g.  you can plot, the average (for every day in a year) of the filling level of the EV battery at every hour of the day, or a heatmap such as the one in Figure 7 in https://doi.org/10.1016/j.enconman.2019.111977 (code [here](https://github.com/martavp/PyPSA-plots/blob/master/scripts/4_plot_charge_discharge_hotmaps.py) )


In short, try to plot results so that you can show that the code is working

martavp replied:
I see now that new figures were added by @ebbekyhl, super!

But still some of my questions regarding the plots 'generators.png' and 'vehiclenb.png' apply

martavp reviewed (changes requested):
Great first step, @s8au , I run using Snakemake and all works fine. 
I included a lot of feedback for this PR. Keep adding commits to address my comments or discuss them below each comment if you don’t agree/understand any of them.

I also created a lot of issues to order the next steps.

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -111,36 +113,92 @@ def transport(string):
+    df_costs.set_index(df_costs.columns[0],append=True,inplace=True)
+    #set parameter/value names as column name, min_count=1 fills not existing values with NaN
+    costs = df_costs.loc[:,"value"].unstack(level=1).groupby("technology").sum(min_count=1)
+    network.add("Store", "battery storage",
```
The "battery storage" should only be added if 'bev_dsm' (demand side management) is activated in the config file
https://github.com/PyPSA/pypsa-eur-sec/blob/e3ffb66267c324b259b4a3db2cf03a3a642044c6/scripts/prepare_sector_network.py#L1508

s8au replied:
Should be solved with commit bcf96f98b34076525d7292b7f6105ddb630ac759

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -111,36 +113,92 @@ def transport(string):
+                capital_cost=costs.at['battery storage', 'investment'],
+                lifetime = costs.at['battery storage', 'lifetime'])
+    network.add("Carrier",'battery')
+    network.add("Link",
```
this link is missing a representation of when the battery cars are parked and connected to the grid so that they can be charged, check @ricnogfer implementation when he added it

martavp replied:
I see this is addressed here: https://github.com/ricnogfer/pypsa-eur-sec_land_transport/pull/24

s8au committed:
Add constraints for EV and ICE
- EV use is represented with two links: battery charger, EV
- constraints ensure consistency of number of EVs and ICEs
- plots for generator use and ration between EVs and ICEs
- rename busses and links for better consistency with PyPSA-Eur

martavp commented:
Thanks for addressing almost all the comments. IMO, this is ready to merge. I suggest you merge this PR and create a new one with other additions. 

Issues that still remain open

[ ] Improving the plots to understand the results, see [my comment below](https://github.com/ricnogfer/pypsa-eur-sec_land_transport/pull/7#discussion_r1208987535)

[ ] The availability of the link for EV charging should be defined as in 
https://github.com/PyPSA/pypsa-eur/blob/80eadcf24aea9b1805d8ebfa66de0422e10f0c04/scripts/prepare_sector_network.py#L1505

[ ] You can use the output of prepare_costs (you can address this in the next PR)

[ ] Next time, you can try to split the additions into several commits that make it easier to review (not a big deal, just a small recommendation)
